utils/pubsub.go

```diff
@@ -14,56 +14,112 @@ type DataEvent struct {
-		go func(data DataEvent, dataChannelSlices DataChannelSlice) {
-			for _, ch := range dataChannelSlices {
-				//Publishing to channels in separate goroutines
-				go func(ch DataChannel, data DataEvent) {
-					ch <- data
-				}(ch, data)
-			}
-		}(DataEvent{Data: data, Topic: topic}, channels)
+	eb.subscriptionsMutex.RLock()
+	defer eb.subscriptionsMutex.RUnlock()
+	eb.lastEventMutex.Lock()
+	defer eb.lastEventMutex.Unlock()
+
+	evt := &DataEvent{Data: data, Topic: topic}
+	if eb.lastEvent == nil {
+		eb.lastEvent = map[string]*DataEvent{}
 	}
-	eb.rm.RUnlock()
-}
+	eb.lastEvent[topic] = evt
 
-func (eb *EventBus) PublishToChannel(channel DataChannel, topic string, data interface{}) {
-	eb.rm.RLock()
-	go func() {
-		channel <- DataEvent{Data: data, Topic: topic}
-	}()
-	eb.rm.RUnlock()
+	if subPublishers, found := eb.subscriptions[topic]; found {
+		for _, subPublisher := range subPublishers {
+			subPublisher.publish(evt)
+		}
+	}
 }
 
 func (eb *EventBus) Subscribe(topic string, ch DataChannel) {
-	eb.rm.Lock()
-	if prev, found := eb.subscribers[topic]; found {
-		eb.subscribers[topic] = append(prev, ch)
+	eb.subscriptionsMutex.Lock()
+	defer eb.subscriptionsMutex.Unlock()
+	eb.lastEventMutex.RLock()
+	defer eb.lastEventMutex.RUnlock()
+
+	newSubPublisher := &subPublisher{channel: ch}
+	if prev, found := eb.subscriptions[topic]; found {
+		eb.subscriptions[topic] = append(prev, newSubPublisher)
 	} else {
-		if eb.subscribers == nil {
-			eb.subscribers = map[string]DataChannelSlice{}
+		if eb.subscriptions == nil {
+			eb.subscriptions = map[string]subPublishers{}
 		}
-		eb.subscribers[topic] = append([]DataChannel{}, ch)
+		eb.subscriptions[topic] = subPublishers{newSubPublisher}
 	}
-	eb.rm.Unlock()
+	if eb.lastEvent[topic] != nil {
+		newSubPublisher.publish(eb.lastEvent[topic])
+	}
+}
+
+// subPublishers is a slice of subPublisher pointers
+type subPublishers []*subPublisher
+
+// subPublisher is responsible to publish events to a single subscription (channel).
+type subPublisher struct {
+	// the channel of the subscription where events are published
+	channel chan DataEvent
+
+	lastValueLock sync.Mutex
+	// the last value waiting to be published to the channel
+	lastValue *DataEvent
+
+	startedLock sync.Mutex
+	// flag indicating whether the publisher's internal goroutine is started or not
+	started bool
+}
+
+// publish sets the publisher's lastValue and starts the
+// internal goroutine if it is not already started
+func (r *subPublisher) publish(data *DataEvent) {
+
+	r.lastValueLock.Lock()
+	// update last value
+	r.lastValue = data
+	r.lastValueLock.Unlock()
+
+	r.startedLock.Lock()
+	// start publish loop if not started
+	if !r.started {
+		go r.startLoop()
+		r.started = true
+	}
+	r.startedLock.Unlock()
+}
+
+// startLoop publishes lastValues to the subscription's channel until there is no other lastValue to publish
+func (r *subPublisher) startLoop() {
+
+	for v := r.nextValue(); v != nil; v = r.nextValue() {
```

Comment:
WDYT of this?

```go
func (r *subPublisher) startLoop() {
	for {
		r.lastValueLock.Lock()
		v := r.lastValue
		r.lastValue = nil
		r.lastValueLock.Unlock()

		if v == nil {
			break
		}

		r.channel <- *v
	}

	r.startedLock.Lock()
	r.started = false
	r.startedLock.Unlock()
}
```

Reply:
I haven't tested it but it looks like it could work.

Reply:
Is it arguably better? I am not really fond of breaks, they kind of mess with the flow

Reply:
I personally find it more readable as I got everything there instead of following yet another method (i.e. `nextValue()`), plus I have no problem with a single `break`, I actually find it very clear especially considering that it is not nested. Up to you.

I pulled your new code and I'm unable to run a test btw, does it happen to you too?

```
=== RUN   TestPubSub_two_consumers_slow_and_fast
started
started
    pubsub_test.go:86: Expected slow consumer to have consumed 5, instead got 2.
```

It hangs like that forever after printing the error.

Reply:
pull again!

Reply:
looking into it

Reply:
Same :thinking: It runs fine without the `-race` though.

Reply:
@atzoum is there a specific reason why these lib doesn't not support contexts? How do we cancel operations and propagate those cancels? How do we cleanup this since this `EventBus` has no `Close()` method?

I'm asking because I might have a solution in mind, I tested it and it doesn't suffer of the blocking op but it would make more sense to have support for op cancelation and cleanups first.

Reply:
no specific reason, but how is this relevant for the tests?

Reply:
I haven't found the reason why the tests block btw, I'll look into it a bit further again tomorrow.

Reply:
it would make sense to have contexts, if it helps try it out :)

Reply:
I will try to fix the test in the meantime1

Reply:
I did this real quick, I'd like to know your opinion: https://github.com/rudderlabs/rudder-server/pull/1813
I can try to add contexts and graceful termination too.

Reply:
it is important not to block the publish method

Reply:
@fracasula there was an issue with the tests. The current implementation has an important characteristic that the tests ignored, which I can explain tomorrow on a call

Reply:
> I did this real quick, I'd like to know your opinion: #1813 I can try to add contexts and graceful termination too.

@fracasula adopted!